File: python/isBalanced.py
import time
"""
By this logic, we say a sequence of brackets is balanced if the 
following conditions are met:
1. It contains no unmatched brackets.
2. The subset of brackets enclosed within the confines of a matched pair of 
brackets is also a matched pair of brackets.
Given  strings of brackets, determine whether each sequence of 
brackets is balanced. If a string is balanced, return YES. 
Otherwise, return NO.
"""
"""
1. create an object where key value pair are opening and closing brackets
2. create an stack (cache) to push opening brackets
3. loop through string
    a. if opening bracket push to cache
    b. if not closing bracket pop from cache 
    c. if poped bracket not in value pair return `NO`
"""
class Stack:

        def __init__(self):
            self.storage = []

        def push(self, value):
            self.storage.append(value)

        def pop(self):
        # if stack is empty return None
            if not self.storage:
                return None

            return self.storage.pop()

        def length(self):
            return len(self.storage)

def isBalanced(s):
    stack = Stack()
    brackets_pairs = {")": "(", "]": "[", "}": "{"}

    for bracket in s:
        if bracket in '({[':
            stack.push(bracket)
        else:
            open_brace = stack.pop()
            if open_brace is not brackets_pairs[bracket]:
                return "NO"
    if stack.length():
        return "NO"
    return "YES"

""" 
throws runtime error if stack data structure
is not used
"""
def isBalanced_2(s):

    brackets_pairs = {")": "(", "]": "[", "}": "{"}
    cache = []
    for bracket in s:
        if bracket in '({[':
            cache.append(bracket)
        else:
            open_brace = cache.pop()
            if open_brace is not brackets_pairs[bracket]:
                return "NO"
    if len(cache):
        return "NO"
    return "YES"

start_time = time.time()
isBalanced('{[()]}')
isBalanced('{[(])}')
isBalanced('{(([])[])[]}[]')
isBalanced('{(([])[])[]}')
end_time = time.time()
print(f"Stack runtime: {end_time - start_time} seconds")

start_time = time.time()
isBalanced_2('{[()]}')
isBalanced_2('{[(])}')
isBalanced_2('{(([])[])[]}[]')
isBalanced_2('{(([])[])[]}')
end_time = time.time()
print(f"List runtime: {end_time - start_time} seconds")

File: javascript/isBalanced.js
function isBalanced(s) {
    start_timer
    const bracketArr = s.split('')
     const cacheBracket = {
         '{':'}',
         '[': ']',
         '(': ')'
     }
     const arrStore = []
     for (let i = 0; i< bracketArr.length; i++) {
         if(bracketArr[i] === '{' || bracketArr[i] === '[' || bracketArr[i] === '(') {
             arrStore.push(bracketArr[i])
         } else {
             const oppBracket = arrStore.pop()
             if(bracketArr[i] !== cacheBracket[oppBracket] ) {
                return 'NO'
             }
         }
     }
     if(arrStore.length) {
         return 'NO'
     }
     return 'YES'
    }
const start_timer = new Date().getTime().toString()
const end_timer = new Date().getTime().toString()
console.log(start_timer)
console.log(isBalanced('{[()]}'))
console.log(isBalanced('{[(])}'))
console.log(isBalanced('{(([])[])[]}[]'))
console.log(isBalanced('{(([])[])[]}'))
console.log(end_timer)
console.log(end_timer-start_timer)



function balancedBrackets(string) {
    // create an bracketArray from sting of brackets
 const bracketArray = string.split('');
// create a cache of all open bracket as keys and closing brackets as values
const cache = {
    '{':'}',
    '[': ']',
    '(': ')',
    '|': '|'
};
// loop through bracketArr and store opening bracket in an openBracket
const openBracket = [];
 for (let i = 0; i < bracketArray.length; i++) {
     if (bracketArray[i] === '{' || bracketArray[i] === '[' || bracketArray[i] === '(') {
         openBracket.push(bracketArray[i]);
     } else if (bracketArray[i] === '|') {
         if (openBracket[openBracket.length - 1] === '|') {
            openBracket.pop()
         }
         openBracket.push(bracketArray[i]); 
     } 
     else if (bracketArray[i] === '}' || bracketArray[i] === ']' || bracketArray[i] === ')') {
         // if not opening bracket pop and compare last push open bracket
         if (openBracket[openBracket.length - 1] === '|') {
            openBracket.pop()
         }
         const closeBracket = openBracket.pop();
         console.log(openBracket)
        if (bracketArray[i] !== cache[closeBracket]) {
            return 0;
        }
     }
 }